atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/ReloadingCloseableContainer.java

```diff
@@ -0,0 +1,145 @@
+ *
+ * There is no guarantee that a resource provided via {@link #get} will <i>not</i> be closed whilst in use.
+ * Consumers should be resilient to any uses after the resource has been closed, and where necessary, retry.
+ *
+ * After {@link #close()} has returned, no further resources will be created, and all managed resources will be
+ * closed.
+ */
+public final class ReloadingCloseableContainer<T extends AutoCloseable> implements AutoCloseable {
+    private static final SafeLogger log = SafeLoggerFactory.get(ReloadingCloseableContainer.class);
+
+    private final AtomicReference<Optional<T>> currentResource;
+    private final Refreshable<T> refreshableResource;
+    private final Disposable refreshableSubscriptionDisposable;
+    private final ReadWriteLock isClosedLock;
+
+    @GuardedBy("isClosedLock")
+    private volatile boolean isClosed;
+
+    private <K> ReloadingCloseableContainer(Refreshable<K> refreshableParameter, Function<K, T> factory) {
+        this.isClosed = false;
+        this.isClosedLock = new ReentrantReadWriteLock(true);
+        this.currentResource = new AtomicReference<>(Optional.empty());
+        this.refreshableResource = refreshableParameter.map(factoryArg -> createNewResource(factoryArg, factory));
+
+        this.refreshableSubscriptionDisposable = refreshableResource.subscribe(resource -> {
+            Optional<T> maybeResourceToClose = currentResource.getAndSet(Optional.of(resource));
+            maybeResourceToClose.ifPresent(this::shutdownResource);
+        });
+    }
+
+    public static <T extends AutoCloseable, K> ReloadingCloseableContainer<T> of(
+            Refreshable<K> refreshableParameter, Function<K, T> factory) {
+        return new ReloadingCloseableContainer<>(refreshableParameter, factory);
+    }
+
+    private <K> T createNewResource(K factoryArg, Function<K, T> factory) {
+        return runIfNotClosed(
+                () -> factory.apply(factoryArg), "Attempted to create a new resource after the container was closed.");
+    }
+
+    /**
+     * A lock is taken out to ensure no new resources are created after retrieving the current stored
+     * resource to close. By doing so, we avoid closing a cluster and subsequently creating a new one that is
+     * never closed.
+     */
+    @Override
+    public void close() {
+        runWithIsClosedWriteLock(() -> {
+            isClosed = true;
+            refreshableSubscriptionDisposable.dispose();
+            currentResource.get().ifPresent(this::shutdownResource);
+        });
+    }
+
+    /**
+     * Gets the latest resource that reflects any changes in the refreshable, provided {@link #close()} has not
+     * been called.
+     *
+     * The resource returned will be closed after {@link #close} is called, or the refreshable is refreshed, even
+     * if the resource is in active use.
+     *
+     * @throws SafeIllegalStateException if the container was closed prior to getting a resource.
```

Comment:
Will note this is a behaviour change in theory - the container now throws if the container is closed, whereas it didn't before.
In practice, this just makes the exception happen earlier - in the 2 places this will be used, interacting with the closed object will also throw an exception anyway.